test/common/upstream/health_checker_impl_test.cc

```diff
@@ -1452,6 +1494,860 @@ TEST_F(RedisHealthCheckerImplTest, AllDontReuseConnection) {
+    auto config = createGrpcHealthCheckConfig();
+    config.mutable_reuse_connection()->set_value(false);
+    health_checker_.reset(
+        new TestGrpcHealthCheckerImpl(*cluster_, config, dispatcher_, runtime_, random_));
+    health_checker_->addHostCheckCompleteCb([this](HostSharedPtr host, bool changed_state) -> void {
+      onHostStatus(host, changed_state);
+    });
+  }
+
+  void expectSessionCreate() {
+    // Expectations are in LIFO order.
+    TestSessionPtr new_test_session(new TestSession());
+    test_sessions_.emplace_back(std::move(new_test_session));
+    TestSession& test_session = *test_sessions_.back();
+    test_session.timeout_timer_ = new Event::MockTimer(&dispatcher_);
+    test_session.interval_timer_ = new Event::MockTimer(&dispatcher_);
+    expectClientCreate(test_sessions_.size() - 1);
+  }
+
+  void expectClientCreate(size_t index) {
+    TestSession& test_session = *test_sessions_[index];
+    test_session.codec_ = new NiceMock<Http::MockClientConnection>();
+    test_session.client_connection_ = new NiceMock<Network::MockClientConnection>();
+    connection_index_.push_back(index);
+    codec_index_.push_back(index);
+
+    EXPECT_CALL(dispatcher_, createClientConnection_(_, _, _))
+        .Times(testing::AnyNumber())
+        .WillRepeatedly(InvokeWithoutArgs([&]() -> Network::ClientConnection* {
+          uint32_t index = connection_index_.front();
+          connection_index_.pop_front();
+          return test_sessions_[index]->client_connection_;
+        }));
+    EXPECT_CALL(*health_checker_, createCodecClient_(_))
+        .WillRepeatedly(
+            Invoke([&](Upstream::Host::CreateConnectionData& conn_data) -> Http::CodecClient* {
+              uint32_t index = codec_index_.front();
+              codec_index_.pop_front();
+              TestSession& test_session = *test_sessions_[index];
+              test_session.codec_client_ = new CodecClientForTest(
+                  std::move(conn_data.connection_), test_session.codec_, nullptr, nullptr);
+              return test_session.codec_client_;
+            }));
+  }
+
+  void expectStreamCreate(size_t index) {
+    test_sessions_[index]->request_encoder_.stream_.callbacks_.clear();
+    EXPECT_CALL(*test_sessions_[index]->codec_, newStream(_))
+        .WillOnce(DoAll(SaveArgAddress(&test_sessions_[index]->stream_response_callbacks_),
+                        ReturnRef(test_sessions_[index]->request_encoder_)));
+  }
+
+  void respondServiceStatus(size_t index,
+                            grpc::health::v1::HealthCheckResponse::ServingStatus status) {
+    respondResponseSpec(index,
+                        ResponseSpec{{{":status", "200"}, {"content-type", "application/grpc"}},
+                                     {ResponseSpec::validChunk(status)},
+                                     {{"grpc-status", "0"}}});
+  }
+
+  void respondResponseSpec(size_t index, ResponseSpec&& spec) {
+    bool trailers_empty = spec.trailers.size() == 0U;
```

Comment:
Nit: can you constify the local tmps where they are immutable everywhere in the tests? Thanks.

Reply:
done